UnitySDK/Assets/ML-Agents/Scripts/InferenceBrain/TensorGenerator.cs

```diff
@@ -0,0 +1,240 @@
+    /// integer array of size 1 containing the batch size.
+    /// </summary>
+    public class BatchSizeGenerator : TensorGenerator
+    {
+        public void Execute(Tensor tensor, int batchSize, Dictionary<Agent, AgentInfo> agentInfo)
+        {
+            tensor.Data = new int[] {batchSize};
+        }
+    }
+
+    /// <summary>
+    /// Generates the Tensor corresponding to the SequenceLength input : Will be a one
+    /// dimensional integer array of size 1 containing 1.
+    /// Note : the sequence length is always one since recurrent networks only predicts for
+    /// one step at the time.
+    /// </summary>
+    public class SequenceLengthGenerator : TensorGenerator
+    {
+        public void Execute(Tensor tensor, int batchSize, Dictionary<Agent, AgentInfo> agentInfo)
+        {
+            tensor.Data = new int[] {1};
+        }
+    }
+
+    /// <summary>
+    /// Generates the Tensor corresponding to the VectorObservation input : Will be a two
+    /// dimensional float array of dimension [batchSize x vectorObservationSize].
+    /// It will use the Vector Observation data contained in the agentInfo to fill the data
+    /// of the tensor.
+    /// </summary>
+    public class VectorObservationGenerator : TensorGenerator
+    {
+        public void Execute(Tensor tensor, int batchSize, Dictionary<Agent, AgentInfo> agentInfo)
+        {
+            tensor.Shape[0] = batchSize;
+            var vecObsSizeT = tensor.Shape[1];
+            tensor.Data = new float[batchSize, vecObsSizeT];
+            var agentIndex = 0;
+            foreach (var agent in agentInfo.Keys)
+            {
+                var vectorObs = agentInfo[agent].stackedVectorObservation;
+                for (var j = 0; j < vecObsSizeT; j++)
+                {
+                    tensor.Data.SetValue(vectorObs[j], new int[2] {agentIndex, j});
+                }
+                agentIndex++;
+            }
+        }
+    }
+
+    /// <summary>
+    /// Generates the Tensor corresponding to the Recurrent input : Will be a two
+    /// dimensional float array of dimension [batchSize x memorySize].
+    /// It will use the Memory data contained in the agentInfo to fill the data
+    /// of the tensor.
+    /// </summary>
+    public class RecurrentInputGenerator : TensorGenerator
+    {
+        public void Execute(Tensor tensor, int batchSize, Dictionary<Agent, AgentInfo> agentInfo)
+        {
+            tensor.Shape[0] = batchSize;
+            var memorySize = tensor.Shape[1];
+            tensor.Data = new float[batchSize, memorySize];
+            var agentIndex = 0;
+            foreach (var agent in agentInfo.Keys)
+            {
+                var memory = agentInfo[agent].memories;
+                for (var j = 0; j < memorySize; j++)
+                {
+                    if (memory == null)
+                    {
+                        break;
+                    }
+
+                    if (j >= memory.Count)
```

Comment:
remove this and modify the end-condition of the for-loop to be the min of memorySize and memory.Count